### ISSUE TITLE
RFC: Amplify UI Component Library Refactor

### ISSUE THREAD

**Issue by jordanranz** [aws-amplify/amplify-js#3279]
This issue is a Request For Comments (RFC).  It is intended to elicit community feedback regarding a proposed change to the Amplify JS UI components.  Please feel free to post comments or questions here.

## Current state of Amplify UI Components

Amplify JS currently provides a library of UI components intended to help developers leverage Amplify categories such as API, Auth, Storage, Interactions and XR.  These UI components are implemented in different modules based on a particular frontend framework.  Currently these frameworks include:

* React: [aws-amplify-react](https://github.com/aws-amplify/amplify-js/tree/master/packages/aws-amplify-react)
* React Native: [aws-amplify-react-native](https://github.com/aws-amplify/amplify-js/tree/master/packages/aws-amplify-vue)
* Angular/Ionic: [aws-amplify-angular](https://github.com/aws-amplify/amplify-js/tree/master/packages/aws-amplify-angular)
* Vue: [aws-amplify-vue](https://github.com/aws-amplify/amplify-js/tree/master/packages/aws-amplify-vue)

The components provided by these modules are implemented in each respective JavaScript framework with duplicated logic. UI components are loosely styled using the Amplify UI style guide with some shared and some duplicated CSS. UI components are customizable but this functionality is limited and lacks parity between frameworks.

## The Problem

This framework-specific approach has led to a variety of maintenance and extensibility challenges, including but not limited to:

* Lack of feature-parity across frameworks
* Duplication of code logic and developer efforts
* Divergent design patterns
* Ramp up for learning and maintaining multiple frameworks
* Inconsistent interaction and styling across frameworks
* Decreased developer velocity in creating new components
* Limited customizability of components

### Overview

Currently, we plan to implement an Amplify UI design system using [Web Components](https://developer.mozilla.org/en-US/docs/Web/Web_Components) generated by [StencilJS](https://stenciljs.com/). We plan to refactor the current framework-specific component logic to alternatively use the logic implemented inside of web components from this design system. 

Additionally, we will utilize CSS-in-JS to create a more flexible customization and theming strategy for components. We are evaluating the use of [Emotion](https://emotion.sh/docs/introduction) to build this strategy into our web components.

This approach has a number of benefits associated with it, including but not limited to:

* Reduced code base
* Faster iteration on enhancements, new features and new components
* Greater accessibility for Amplify users working with frameworks other than React, Angular, Ionic, Vue, or those using no framework at all
* More flexible and consistent styling pattern
* Improved developer experience and code quality

### Design Requirements

* Components should be useable within our currently supported browser-based frameworks (React, Angular, Vue) with little-to-no friction for developers using Amplify JS.  This could conceivably entail the creation of framework-specific wrappers for the components, but at this time it is thought that Stencil generated web components should suffice.
* Components must be functional within other web components, UI frameworks, and plain HTML/JS pages.
* This project should offer feature parity with the sum of functionality between our existing UI packages; in other words, developers currently operating within a given framework should not find that they have lost capabilities upon migration.
* Components should be [responsive](https://developer.mozilla.org/de/docs/Web_Development/Mobile/Responsive_design).
* Components should be customizable and themeable.  For example, users of an 'Authenticator' component should be able to define their 'Sign-In' child component which uses their own markup and styling, but which seamless integrates with the out-of-the-box behavior of the Authenticator.
* All efforts should be taken to maintain as small a [bundle size](https://material-ui.com/guides/minimizing-bundle-size/) as possible.
* Components should be Typescript-compatible with well-defined and understandable typings.
* Components must be compatible with [AWS supported browsers](https://aws.amazon.com/premiumsupport/knowledge-center/browsers-management-console/).
* SSR (Server Side Rendering) of Amplify UI components should be supported

### Customizable components

* Add attribute for isSignInDisplayed to config for amplify-sign-up (https://github.com/aws-amplify/amplify-js/issues/3216)
* Feature request: PhotoPicker should allow setting id (https://github.com/aws-amplify/amplify-js/issues/3196)
* Direct routeLink to subforms of <amplify-authenticator> (https://github.com/aws-amplify/amplify-js/issues/3191)
* Vue Authenticator Component missing theme integration (https://github.com/aws-amplify/amplify-js/issues/3068) 
* Forgot Password? text should be clickable (https://github.com/aws-amplify/amplify-js/issues/3046)
* Allow theming for select elements (https://github.com/aws-amplify/amplify-js/issues/2994)
* Angular theme suport (https://github.com/aws-amplify/amplify-js/issues/2750)
* Customize amplify-react Greetings className not update to Nav #help wanted (https://github.com/aws-amplify/amplify-js/issues/2672)
* [Vue] How to safely override Amplify built-in components via CSS? (https://github.com/aws-amplify/amplify-js/issues/2471)
* Provide a function to customize error message (https://github.com/aws-amplify/amplify-js/issues/2104)
* Customize AWS Amplify Authenticator UI for Angular (https://github.com/aws-amplify/amplify-js/issues/2075)
* Theme not available on amplify-angular (https://github.com/aws-amplify/amplify-js/issues/1755)
* how can we style/scale the image displayed by <S3Image> component for React ? (https://github.com/aws-amplify/amplify-js/issues/1655)
* React customize-withauthenticator doesnt work on v2.0.0-2.0.1 (https://github.com/aws-amplify/amplify-js/issues/1603)
* Custom UI withAuthenticator (https://github.com/aws-amplify/amplify-js/issues/1436)
* Increase theme-bility in Ionic 4 (https://github.com/aws-amplify/amplify-js/issues/1358)
* Passing down className prop in S3Image component (https://github.com/aws-amplify/amplify-js/issues/1272)
* Is it possible to customize default Sign Up, Sign In screens of aws-amplify-react-native package? (https://github.com/aws-amplify/amplify-js/issues/1031)
* RFC: react component HOCs allowing for bringing your own component (https://github.com/aws-amplify/amplify-js/issues/940)
* RFC: Authenticator render props (https://github.com/aws-amplify/amplify-js/issues/474)
* Can I validate user input w/o subsequent signup? Only validation (https://github.com/aws-amplify/amplify-js/issues/1054)

### Component improvements

* Angular authenticator sign in fill login form (https://github.com/aws-amplify/amplify-js/issues/3050)
* amplify-authenticator angular adding rememberme (https://github.com/aws-amplify/amplify-js/issues/3010)
* aws-amplify-react: User cannot be confirm. Current status is CONFIRMED (https://github.com/aws-amplify/amplify-js/issues/2797)
* Federated sign-in buttons have broken styles. (https://github.com/aws-amplify/amplify-js/issues/2752)
* How to disable Amplify from loading its Auth CSS theme file in a React app? (https://github.com/aws-amplify/amplify-js/issues/2691)
* React Native Amplify UI: Phone Number Country Code is forced to be a Picker (https://github.com/aws-amplify/amplify-js/issues/2632)
* Auto signin after confirm email (https://github.com/aws-amplify/amplify-js/issues/2562)
* [Vue] Auth, password not contained within a form element (https://github.com/aws-amplify/amplify-js/issues/2479)
* Could not find a declaration file for module '@aws-amplify/ui' (https://github.com/aws-amplify/amplify-js/issues/2460)
* <amplify-authenticator> password weakness warning is unclear (https://github.com/aws-amplify/amplify-js/issues/2416)
* Allow null as default for countryCode in signup components (https://github.com/aws-amplify/amplify-js/issues/2327)
* Add loading indicator to Authenticator components (https://github.com/aws-amplify/amplify-js/issues/2311)
* importing 'aws-amplify-react', ERROR: Uncaught SyntaxError: Unexpected token : (https://github.com/aws-amplify/amplify-js/issues/2230)
* Make React Authenticator CSS class names customization friendly (https://github.com/aws-amplify/amplify-js/issues/2161)
* Getting error when implement to nextJs (https://github.com/aws-amplify/amplify-js/issues/1963)
* authentication ui (https://github.com/aws-amplify/amplify-js/issues/404)
* S3Image component ignores preview="hidden" if path prop specified (https://github.com/aws-amplify/amplify-js/issues/218)
* aws-amplify-react-native UI Components (https://github.com/aws-amplify/amplify-js/issues/2725)
* Add Auth to React context using Provider pattern (https://github.com/aws-amplify/amplify-js/issues/3090)
* Verify Contact flow for logged-in user (https://github.com/aws-amplify/amplify-js/issues/4503)

Related milestone: https://github.com/aws-amplify/amplify-js/milestone/13

**Comment by ajhool** [comment]
Note: all experience has been with the React components.

I'm glad to see that the team will refactor these components. 

1. Please try to add SSR compatibility into any new additions to Amplify. It does not need to actually render on the server, but please ensure that it does not break SSR. Currently the React Authenticator component breaks SSR and requires fragile workarounds

2. The documentation is currently too sparse and unorganized. The documentation is all example-based and I've had trouble finding an explanation a simple, plain reference for the various props. The Authenticator props and design choices, in particular, are strange and unintuitive, so that component has been frustrating to work with despite being (arguably) the most important UI in any application.

3. Typescript instead of javascript

4. The Authenticator component requires passing in Components to the "hide" prop (eg. `<Greetings />, <Loading />`). It would be nice to just be able to pass in strings, (eg. ["loading", "greetings"]).

**Comment by jordanranz** [comment]
Hey @ajhool,

Thank you for the feedback. 

1. Creating a smooth interaction for SSR is one of design requirements for this project. I must have missed adding it to the list above. Using both StencilJS and Emotion should help us achieve this goal.

2. Documentation will definitely be refactored over the course of this project. One big improvement we are looking to make comes with introducing the use of [Storybook](https://storybook.js.org/) to our documentation. This will provide a lot more clarity to a user on how to use the component API and also even allow them to test the functionality our components.

3. There is a design requirement to make this happen: "Components should be Typescript-compatible with well-defined and understandable typings". StencilJS generated web components are [typed](https://stenciljs.com/docs/typed-components/) out-of-the-box, so this should be fairly straightforward with the current development plan.

4. I will make note of this feedback and refer to it when building the new Authenticator web component. Passing a string array is a good idea but maintaining a string to component mapping might not be the most scalable option. Definitely something to look into more.

5. Noted. This is more of a library or organizational piece of feedback but we will definitely take this into account moving forward. Also FYI there is a PR for that specific callout: https://github.com/aws-amplify/amplify-js/pull/3103 :)

6. Good comments and feedback. We can evaluate this when creating the new Authenticator component. Maybe have this order configurable where you can choose the order of your Authenticator flow (i.e Sign-up first, with a "I already have an account" link or Sign-in first, with an account sign up link).

**Comment by ajhool** [comment]
Thanks for the response and I look forward to contributing towards Storybook integration and Typescript compatibility, these are excellent tools! Would it be helpful if I created a Storybook integration for the existing UI components?

**Comment by jordanranz** [comment]
We would greatly appreciate the contribution of Storybook integration with our existing UI components and accept the use of Typescript to build this! 

If this is something you would like to contribute, you can open a new issue to track this feature request/implementation and reference this RFC.

**Comment by clementAC** [comment]
Same note as @ajhool 

Hi,

I agree with ajhool  on all the points. 

The doc is really great for a beginner with lot of exemples but once you try to make more complex uses cases there are some important points and props that are not documented. And searching in source code is quite annoying. Storybook or a similar product could be a good solution

I also agree on the amplify lib development "break-neck" thing. Sometimes I'm completely wondering why a feature is not implemented whereas it's  a typical use case for a website. For instance, user administration, choose on wich s3 bucket I want to upload my files (default one or another one) and lot more and i'm not talking about visual components.
I know that amplify is created for simplifying aws services access and adoption (and it manage to do it) but if the classical use cases are not taken into account then the point is completelly missed and sometimes amplify is making the service access more "frictional".

Everytime amplify is updated I'm wondering what have changed. With this new approach it could be an excellent idea to implement a changelog system. It will sure avoid some side effects du to upgrading the library in the future.

During the last month amplify have changed a lot and I think this proposition of refactoring is a great news for amplify developers and users.

**Comment by ajhool** [comment]
Have started a storybook integration here: https://github.com/ajhool/amplify-js . If anybody is available from the Amplify team for a quick realtime chat this week to help me get oriented, I'd really appreciate it. After that, it should be ("should") relatively straightforward to build stories for each component

**Comment by buggy** [comment]
@jordanranz What's the timeline on this like? I'm wondering if it's worth continuing with #2436?

**Comment by jordanranz** [comment]
@buggy, Shooting for end of the year release. I think your PR is definitely still worth it. I do not plan on removing `aws-amplify-react` and a lot of the work that has been done in that PR will be reused as we will continue to use those React components as wrappers.

**Comment by jkeys-ecg-nmsu** [comment]
@jordanranz @ajhool I think the tag "pending-triage" says it all. I'm asking for features all the time, as is much of the Amplify dev community. (I've also not yet contributed to the codebase directly. =/ ) My company is in its start-up cycle trying to make stuff that doesn't exist yet, so we tend to work around what isn't stable or working instead of relying on it being fixed in the next iteration.

Perhaps Amplify could break the traditional AWS mold by releasing a roadmap? This might also encourage more contributions from the community.

**Comment by jordanranz** [comment]
Thanks for the feedback @jkeys-ecg-nmsu. I think presenting more RFCs like this one is a step in the right direction for exposing a more public roadmap. We have also discussed presenting more granular roadmap items to the community but haven't settled on the best way to do this.

This being said, I don't want to stray too far off topic from the RFC being presented. I would love to hear more feedback about specific component customizations that Amplify UI users need or would like to be improved.

**Comment by homerjam** [comment]
Please focus on auth - the rest are nice to have, non-essential components. Look at Auth0 for a model of how slick you can make auth.

Please add a note to the documentation about the lack of feature parity between the current frameworks and the timeline for the refactor.

I personally don't care for customisation too much - I think these components should focus on being reliable, universal and accessible. I'd like to be able to use drop them in whilst prototyping and for them to 'just work'.

**Comment by shicholas** [comment]
hey @homerjam, shameless shout-out - I took some time and solved Auth w/ Cognito User Pools & AWS AppSync for React apps if you're interested. 

If you are, please check out https://design.1337lawyers.com/?path=/story/authentication--setup and lmk how I can improve it. It uses Formik to wrap Auth calls, and exports a `CurrentUserConsumer` that uses the React context API to give you information about your current Cognito user when/where you need it.

**Comment by homerjam** [comment]
Thanks @shicholas! I made progress also, I eventually managed to get persisted federated sign in working with Nuxt (SSR) using a combination of Amplify, AppSync and deprecated Cognito libraries. Though having been through all that we're now dropping Amplify in favour of Prisma which I hope will be a better experience.

**Comment by shicholas** [comment]
nice sounds good. haha I hope Vue & React converge one day. There's good ideas in both and it'd be cool if it was all the same.

**Comment by cliffordh** [comment]
@homerjam what were your reasons for going with Prisma

**Comment by undefobj** [comment]
@cliffordh @shicholas @homerjam - This issue is for the team to get feedback on the RFC to improve the UI components for Amplify. If you have unrelated concerns/feedback/questions could you please move them to another issue or forum? Thanks much.

**Comment by yareyaredesuyo** [comment]
Yeah,  Signup/Login UI customization is major problem (this maybe related to Cognito team?). It maybe great to implement Auth UI like UI Components.

The rest of things is trivial.

**Comment by ajhool** [comment]
Issue not stale, obviously

**Comment by mbbillz** [comment]
Semi related to #3358 & #3348 - it would be nice to be able to optionally exclude styles. When using the Authenticator component (aws-amplify-react), the styles that get included (which I do not use as am using custom components) take up nearly half of my final CSS.

**Comment by j0b0sapi3n** [comment]
@mbbillz I have this exact same problem.

**Comment by bruqui** [comment]
I wish I was looking into this framework when this RFC came out; however, from my experience so far, which is fairly new and trying to figure this library out... instead of trying to build out all these forms for all the scenarios in React which has huge limitations such as wanting to use Material Design styled components instead of what's provided and having to build out an entire component with all the functionality just to replace the SignUp component for instance, why not go with an AuthProvider component where things like signUp, signIn, etc. functions along with other needed values from the provider's state can just be pulled out of the provider with useContext. This way only one component needs to be managed and documented. This leaves the rest of the UI for styling, special needs in functionality, etc. left to the user's discretion.

**Comment by mbbillz** [comment]
@bruqui's solution sounds really neat!

**Comment by bruqui** [comment]
It's not far off from the combo of [React Apollo](https://www.apollographql.com/docs/react/api/react-apollo/) and [@apollo/react-hooks](https://www.apollographql.com/docs/react/api/react-hooks/) do really. In fact, following that pattern all together would probably be the most ideal by also providing different kinds of hooks that seem to make the most sense for the provider component.

**Comment by michelschmit** [comment]
Hi, following the discussion here : https://github.com/aws-amplify/amplify-js/issues/4515#issuecomment-575290910

can you add this to your things to do on amplify-ui : 
currently you use `var()` to handle colors but the css variables aren't supported by IE11.

(@ericclemmons, tell me if it's enough or if i have to open a specific issue and link it here. )

Thanks !

**Comment by Rafcin** [comment]
This probably isn't right but as of now I think in terms of signup/signin forms aren't all bad. The default auth ui sucks but you can easily make your own! I Found using formik works well for auth forms, it's simple where you can make checks for basic things like "is it an email" or "is it a zipcode", but what I think would be nice is a way to check something like isValidPassword() for when you login because while the auth form handles all this, not all people want to use it. In my case I use filepond for uploading files, and I can't do that with the auth form, so these tiny details would be nice.

Or add a component that's highly customizable. 

A side suggestion would be to change the community channel Glitter to something like Slack. One major problem is if you want to search something it's almost impossible on Glitter. Some things on the wiki are not up to date or complex and when someone explains it it's impossible to find and or the chat is dead and no one really has a solution. I think Slack or some other program would be nice and it would allow us to easily communicate, possibly track issues from Slack or discuss problems and solutions with  AWS devs and the community.

**Comment by mayteio** [comment]
Is there a list of issues related to this RFC, or is it still in discussion? I'd love to start contributing and it would be awesome if related issues could be triaged with a label for it and good-first-issues too.

**Comment by ProdigySim** [comment]
We used amplify-react, then amplify-react-native, and then built a custom UI for the next project.

There are so few AmplifyJS API calls needed for each auth component that I'd rather see good documentation on how to set up an "Authenticator" yourself. Handling auth state changes + specific API calls to do Signin, forgot password, etc.

Building a signin form is not hard. Handling Amplify's auth flows properly is. A react context object or an angular service with the relevant APIs exposed would go a lot further than a UI in my opinion. Help us model our UI in the correct form, but don't make us try to customize a premade one.

**Comment by jordanranz** [comment]
# Amplify UI Component (Preview)

Hello all,

Wanted to provide an update on the status for this project and give an overview of the phases we are planning on releasing in the near future. As mentioned above in [this RFC](https://github.com/aws-amplify/amplify-js/issues/3279#issue-444158769) the problems we wanted to solve for this project included:

* Lack of feature-parity across frameworks
* Duplication of code logic and developer efforts
* Divergent design patterns
* Ramp up for learning and maintaining multiple frameworks
* Inconsistent interaction and styling across frameworks
* Decreased developer velocity in creating new components
* Limited customizability of components

### Phase 1

In **Phase 1** of this project we will seek to solve or at least alleviate all of the problems listed above. To achieve this we will be releasing a brand new web component library and new libraries for each of the frameworks that we currently support. The framework libraries will include React, Angular, and Vue and these will be generated based off of the code in the [core web components library](https://github.com/aws-amplify/amplify-js/tree/ui-components/master/packages/amplify-ui-components). The original component libraries offered in the current release of amplify-js will go into maintenance mode and we will no longer be adding features to these libraries. For customers using the original React, Angular, or Vue component libraries, we will be providing migration guides for those looking to migrate to using the new components. **Phase 1** is complete when we are at feature parity with the original UI component libraries.

[Here](https://github.com/aws-amplify/amplify-js/tree/ui-components/master/packages/amplify-ui-components#frameworks) is a breakdown of the new libraries being added:

| Framework          | Package                                                                                  | Version                                                                                                                                        |
| ------------------ | ---------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------- |
| **React**          | [`@aws-amplify/ui-react`](https://www.npmjs.com/package/@aws-amplify/ui-react)           | [![version](https://img.shields.io/npm/v/@aws-amplify/ui-react/ui-preview.svg)](https://www.npmjs.com/package/@aws-amplify/ui-react)           |
| **Angular**        | [`@aws-amplify/ui-angular`](https://www.npmjs.com/package/@aws-amplify/ui-angular)       | [![version](https://img.shields.io/npm/v/@aws-amplify/ui-angular/ui-preview.svg)](https://www.npmjs.com/package/@aws-amplify/ui-angular)       |
| **Vue**            | [`@aws-amplify/ui-vue`](https://www.npmjs.com/package/@aws-amplify/ui-vue)               | [![version](https://img.shields.io/npm/v/@aws-amplify/ui-vue/ui-preview.svg)](https://www.npmjs.com/package/@aws-amplify/ui-vue)               |
| **Web Components** | [`@aws-amplify/ui-components`](https://www.npmjs.com/package/@aws-amplify/ui-components) | [![version](https://img.shields.io/npm/v/@aws-amplify/ui-components/ui-preview.svg)](https://www.npmjs.com/package/@aws-amplify/ui-components) |


For feedback and feature requests related to this preview release please [open a new issue](https://github.com/aws-amplify/amplify-js/issues) in our issue queue and add the label: [Amplify UI Components](https://github.com/aws-amplify/amplify-js/labels/Amplify%20UI%20Components). Also, feel free to comment below on this RFC.


### Phase 2

Once our new UI component libraries are at feature parity with the original libraries, we will move on to **Phase 2**. In this phase we will be looking to address the comments and requests presented in this RFC. I really like the ideas presented here especially the one about offering an AuthProvider “renderless” component mentioned in @bruqui‘s comment. Interactive and better documentation as @ajhool mentioned is also something we will be looking to support.

In addition to addressing the comments in this RFC, we will also be doing a retro on the issues and feature requests in our queue [related to UI components](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aopen+is%3Aissue+label%3AUI). [This milestone](https://github.com/aws-amplify/amplify-js/milestone/13) will be used to track the progress for these.

**Phase 2** may also include new components from other Amplify categories like Predictions or API. If anyone would like to contribute ideas to what components you’d like to see added in the future please post it in this RFC or open a new feature request with the [Amplify UI Components](https://github.com/aws-amplify/amplify-js/labels/Amplify%20UI%20Components) label!


## Thank you!

The feedback that you guys give is extremely helpful and gives us a better understanding of what people want us to build. Go ahead and try out the [new component libraries](https://github.com/aws-amplify/amplify-js/tree/ui-components/master/packages/amplify-ui-components#quick-start) and let us know what you think.

**Comment by jordanranz** [comment]
The Phase 1 of the post I made above has been released for the Auth category components!

Here is the blog post:
https://aws.amazon.com/blogs/mobile/amplify-framework-announces-new-rearchitected-ui-component-and-modular-javascript-libraries/

Here is out new preview Docs:
https://docs.amplify.aws/ui

We will now be working on adding additional category support for Storage, API, etc. and then moving on to Phase 2. I am going to close this RFC now. Thank you everyone for the feedback on this Issue!

Cheers 🍻
